Assets/Scripts/EOSTransferTask.cs

```diff
@@ -0,0 +1,305 @@
+* 
+* The above copyright notice and this permission notice shall be included in all
+* copies or substantial portions of the Software.
+* 
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+*/
+
+namespace PlayEveryWare.EpicOnlineServices.Samples
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Epic.OnlineServices;
+    using Epic.OnlineServices.PlayerDataStorage;
+    using UnityEngine;
+
+    public abstract class EOSTransferTask
+    {
+        /// <summary>
+        /// Per EOS SDK documentation, the maximum size for a file is 200MB, or this many bytes.
+        /// If a file transfer is started that exceeds this value, it will likely fail.
+        /// When setting <see cref="TotalSize"/> this value is checked and warned against.
+        /// </summary>
+        public const int FILE_MAX_SIZE_BYTES = 200000000;
+
+        /// <summary>
+        /// This is the most amount of bytes that can be written at once while writing.
+        /// </summary>
+        public const uint MAX_CHUNK_SIZE = 4096;
+
+        /// <summary>
+        /// When an EOSTransferTask is made, it either is an immediate failure or an inner task should be set.
+        /// This keeps a reference to the entire task completion source so that it can complete it when done.
+        /// </summary>
+        public TaskCompletionSource<byte[]> InnerTaskCompletionSource;
+
+        /// <summary>
+        /// Delegate signature for <see cref="OnProgressUpdated"/>.
+        /// </summary>
+        /// <param name="progress">The current percentage completion of the task, by number of bytes to work on.</param>
+        public delegate void OnProgressUpdatedDelegate(float progress);
+
+        /// <summary>
+        /// An optional task that can be utilized to inform the user of file transfer progress.
+        /// </summary>
+        public event OnProgressUpdatedDelegate OnProgressUpdated;
+
+        /// <summary>
+        /// If this task is a download task, this is where the data will be put when the task is complete.
+        /// If this is an upload task, this is where the data should be put.
+        /// </summary>
+        public byte[] Data { get; protected set; } = new byte[] { };
+
+        /// <summary>
+        /// If this task has run to completion, or encountered an error, this ResultCode should be set.
+        /// You can determine that an EOSTransferTask is complete if it has any result code.
+        /// </summary>
+        public Result? ResultCode;
```

Comment:
You should explain in your comments why you are using a `nullable` value, and what it means when this field member is `null`.